File: samples/sample_arm.c
/* Unicorn Emulator Engine */
/* By Nguyen Anh Quynh, 2015 */

/* Sample code to demonstrate how to emulate ARM code */

#include <unicorn/unicorn.h>
#include <string.h>

// code to be emulated
// #define ARM_CODE "\x37\x00\xa0\xe3" // mov r0, #0x37
#define ARM_CODE "\x00\xf0\x20\xe3" // nop
// #define ARM_CODE "\x37\x00\xa0\xe3\x03\x10\x42\xe0" // mov r0, #0x37; sub r1,
// r2, r3
#define THUMB_CODE "\x83\xb0" // sub    sp, #0xc

#define ARM_THUM_COND_CODE                                                     \
    "\x9a\x42\x14\xbf\x68\x22\x4d\x22" // 'cmp r2, r3\nit ne\nmov r2, #0x68\nmov
                                       // r2, #0x4d'

// code to be emulated
#define ARM_CODE_EB                                                            \
    "\xe3\xa0\x00\x37\xe0\x42\x10\x03" // mov r0, #0x37; sub r1, r2, r3
#define THUMB_CODE_EB "\xb0\x83"       // sub    sp, #0xc
// memory address where emulation starts
#define ADDRESS 0x10000

static void hook_block(uc_engine *uc, uint64_t address, uint32_t size,
                       void *user_data)
{
    printf(">>> Tracing basic block at 0x%" PRIx64 ", block size = 0x%x\n",
           address, size);
}

static void hook_code(uc_engine *uc, uint64_t address, uint32_t size,
                      void *user_data)
{
    printf(">>> Tracing instruction at 0x%" PRIx64
           ", instruction size = 0x%x\n",
           address, size);
}

static void test_arm(void)
{
    uc_engine *uc;
    uc_err err;
    uc_hook trace1, trace2;

    int r0 = 0x1234; // R0 register
    int r2 = 0x6789; // R1 register
    int r3 = 0x3333; // R2 register
    int r1;          // R1 register

    printf("Emulate ARM code\n");

    // Initialize emulator in ARM mode
    err = uc_open(UC_ARCH_ARM, UC_MODE_ARM, &uc);
    if (err) {
        printf("Failed on uc_open() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    // map 2MB memory for this emulation
    uc_mem_map(uc, ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

    // write machine code to be emulated to memory
    uc_mem_write(uc, ADDRESS, ARM_CODE, sizeof(ARM_CODE) - 1);

    // initialize machine registers
    uc_reg_write(uc, UC_ARM_REG_R0, &r0);
    uc_reg_write(uc, UC_ARM_REG_R2, &r2);
    uc_reg_write(uc, UC_ARM_REG_R3, &r3);

    // tracing all basic blocks with customized callback
    uc_hook_add(uc, &trace1, UC_HOOK_BLOCK, hook_block, NULL, 1, 0);

    // tracing one instruction at ADDRESS with customized callback
    uc_hook_add(uc, &trace2, UC_HOOK_CODE, hook_code, NULL, ADDRESS, ADDRESS);

    // emulate machine code in infinite time (last param = 0), or when
    // finishing all the code.
    err = uc_emu_start(uc, ADDRESS, ADDRESS + sizeof(ARM_CODE) - 1, 0, 0);
    if (err) {
        printf("Failed on uc_emu_start() with error returned: %u\n", err);
    }

    // now print out some registers
    printf(">>> Emulation done. Below is the CPU context\n");

    uc_reg_read(uc, UC_ARM_REG_R0, &r0);
    uc_reg_read(uc, UC_ARM_REG_R1, &r1);
    printf(">>> R0 = 0x%x\n", r0);
    printf(">>> R1 = 0x%x\n", r1);

    uc_close(uc);
}

static void test_thumb(void)
{
    uc_engine *uc;
    uc_err err;
    uc_hook trace1, trace2;

    int sp = 0x1234; // R0 register

    printf("Emulate THUMB code\n");

    // Initialize emulator in ARM mode
    err = uc_open(UC_ARCH_ARM, UC_MODE_THUMB, &uc);
    if (err) {
        printf("Failed on uc_open() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    // map 2MB memory for this emulation
    uc_mem_map(uc, ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

    // write machine code to be emulated to memory
    uc_mem_write(uc, ADDRESS, THUMB_CODE, sizeof(THUMB_CODE) - 1);

    // initialize machine registers
    uc_reg_write(uc, UC_ARM_REG_SP, &sp);

    // tracing all basic blocks with customized callback
    uc_hook_add(uc, &trace1, UC_HOOK_BLOCK, hook_block, NULL, 1, 0);

    // tracing one instruction at ADDRESS with customized callback
    uc_hook_add(uc, &trace2, UC_HOOK_CODE, hook_code, NULL, ADDRESS, ADDRESS);

    // emulate machine code in infinite time (last param = 0), or when
    // finishing all the code.
    // Note we start at ADDRESS | 1 to indicate THUMB mode.
    err = uc_emu_start(uc, ADDRESS | 1, ADDRESS + sizeof(THUMB_CODE) - 1, 0, 0);
    if (err) {
        printf("Failed on uc_emu_start() with error returned: %u\n", err);
    }

    // now print out some registers
    printf(">>> Emulation done. Below is the CPU context\n");

    uc_reg_read(uc, UC_ARM_REG_SP, &sp);
    printf(">>> SP = 0x%x\n", sp);

    uc_close(uc);
}

static void test_armeb(void)
{
    uc_engine *uc;
    uc_err err;
    uc_hook trace1, trace2;

    int r0 = 0x1234; // R0 register
    int r2 = 0x6789; // R1 register
    int r3 = 0x3333; // R2 register
    int r1;          // R1 register

    printf("Emulate ARM Big-Endian code\n");

    // Initialize emulator in ARM mode
    err = uc_open(UC_ARCH_ARM, UC_MODE_ARM + UC_MODE_BIG_ENDIAN, &uc);
    if (err) {
        printf("Failed on uc_open() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    // map 2MB memory for this emulation
    uc_mem_map(uc, ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

    // write machine code to be emulated to memory
    uc_mem_write(uc, ADDRESS, ARM_CODE_EB, sizeof(ARM_CODE_EB) - 1);

    // initialize machine registers
    uc_reg_write(uc, UC_ARM_REG_R0, &r0);
    uc_reg_write(uc, UC_ARM_REG_R2, &r2);
    uc_reg_write(uc, UC_ARM_REG_R3, &r3);

    // tracing all basic blocks with customized callback
    uc_hook_add(uc, &trace1, UC_HOOK_BLOCK, hook_block, NULL, 1, 0);

    // tracing one instruction at ADDRESS with customized callback
    uc_hook_add(uc, &trace2, UC_HOOK_CODE, hook_code, NULL, ADDRESS, ADDRESS);

    // emulate machine code in infinite time (last param = 0), or when
    // finishing all the code.
    err = uc_emu_start(uc, ADDRESS, ADDRESS + sizeof(ARM_CODE_EB) - 1, 0, 0);
    if (err) {
        printf("Failed on uc_emu_start() with error returned: %u\n", err);
    }

    // now print out some registers
    printf(">>> Emulation done. Below is the CPU context\n");

    uc_reg_read(uc, UC_ARM_REG_R0, &r0);
    uc_reg_read(uc, UC_ARM_REG_R1, &r1);
    printf(">>> R0 = 0x%x\n", r0);
    printf(">>> R1 = 0x%x\n", r1);

    uc_close(uc);
}

static void test_thumbeb(void)
{
    uc_engine *uc;
    uc_err err;
    uc_hook trace1, trace2;

    int sp = 0x1234; // R0 register

    printf("Emulate THUMB Big-Endian code\n");

    // Initialize emulator in ARM mode
    err = uc_open(UC_ARCH_ARM, UC_MODE_THUMB + UC_MODE_BIG_ENDIAN, &uc);
    if (err) {
        printf("Failed on uc_open() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    // map 2MB memory for this emulation
    uc_mem_map(uc, ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

    // write machine code to be emulated to memory
    uc_mem_write(uc, ADDRESS, THUMB_CODE_EB, sizeof(THUMB_CODE_EB) - 1);

    // initialize machine registers
    uc_reg_write(uc, UC_ARM_REG_SP, &sp);

    // tracing all basic blocks with customized callback
    uc_hook_add(uc, &trace1, UC_HOOK_BLOCK, hook_block, NULL, 1, 0);

    // tracing one instruction at ADDRESS with customized callback
    uc_hook_add(uc, &trace2, UC_HOOK_CODE, hook_code, NULL, ADDRESS, ADDRESS);

    // emulate machine code in infinite time (last param = 0), or when
    // finishing all the code.
    // Note we start at ADDRESS | 1 to indicate THUMB mode.
    err = uc_emu_start(uc, ADDRESS | 1, ADDRESS + sizeof(THUMB_CODE_EB) - 1, 0,
                       0);
    if (err) {
        printf("Failed on uc_emu_start() with error returned: %u\n", err);
    }

    // now print out some registers
    printf(">>> Emulation done. Below is the CPU context\n");

    uc_reg_read(uc, UC_ARM_REG_SP, &sp);
    printf(">>> SP = 0x%x\n", sp);

    uc_close(uc);
}

static void test_thumb_mrs(void)
{
    uc_engine *uc;
    uc_err err;
    uc_hook trace1, trace2;

    int pc;

    printf("Emulate THUMB MRS instruction\n");
    // 0xf3ef8014 - mrs r0, control

    // Initialize emulator in ARM mode
    err = uc_open(UC_ARCH_ARM, UC_MODE_THUMB, &uc);
    if (err) {
        printf("Failed on uc_open() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    // Setup the cpu model.
    err = uc_ctl_set_cpu_model(uc, UC_CPU_ARM_CORTEX_M33);
    if (err) {
        printf("Failed on uc_ctl() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    // map 2MB memory for this emulation
    uc_mem_map(uc, ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

    // write machine code to be emulated to memory
    uc_mem_write(uc, ADDRESS, "\xef\xf3\x14\x80", 4);

    // tracing all basic blocks with customized callback
    uc_hook_add(uc, &trace1, UC_HOOK_BLOCK, hook_block, NULL, 1, 0);

    // tracing one instruction at ADDRESS with customized callback
    uc_hook_add(uc, &trace2, UC_HOOK_CODE, hook_code, NULL, ADDRESS, ADDRESS);

    // emulate machine code in infinite time (last param = 0), or when
    // finishing all the code.

    // Note we start at ADDRESS | 1 to indicate THUMB mode.
    err = uc_emu_start(uc, ADDRESS | 1, ADDRESS + 4, 0, 1);
    if (err) {
        printf("Failed on uc_emu_start() with error returned: %u\n", err);
    }

    // now print out some registers
    printf(">>> Emulation done. Below is the CPU context\n");

    uc_reg_read(uc, UC_ARM_REG_PC, &pc);
    printf(">>> PC = 0x%x\n", pc);
    if (pc != ADDRESS + 4) {
        printf("Error, PC was 0x%x, expected was 0x%x.\n", pc, ADDRESS + 4);
    }

    uc_close(uc);
}

static void test_thumb_ite_internal(bool step, uint32_t *r2_out,
                                    uint32_t *r3_out)
{
    uc_engine *uc;
    uc_err err;

    uint32_t sp = 0x1234;
    uint32_t r2 = 0, r3 = 1;

    err = uc_open(UC_ARCH_ARM, UC_MODE_THUMB, &uc);
    if (err) {
        printf("Failed on uc_open() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    uc_mem_map(uc, ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

    uc_mem_write(uc, ADDRESS, ARM_THUM_COND_CODE,
                 sizeof(ARM_THUM_COND_CODE) - 1);

    uc_reg_write(uc, UC_ARM_REG_SP, &sp);

    uc_reg_write(uc, UC_ARM_REG_R2, &r2);
    uc_reg_write(uc, UC_ARM_REG_R3, &r3);

    if (!step) {
        err = uc_emu_start(uc, ADDRESS | 1,
                           ADDRESS + sizeof(ARM_THUM_COND_CODE) - 1, 0, 0);
        if (err) {
            printf("Failed on uc_emu_start() with error returned: %u\n", err);
        }
    } else {
        int i, addr = ADDRESS;
        for (i = 0; i < sizeof(ARM_THUM_COND_CODE) / 2; i++) {
            err = uc_emu_start(uc, addr | 1,
                               ADDRESS + sizeof(ARM_THUM_COND_CODE) - 1, 0, 1);
            if (err) {
                printf("Failed on uc_emu_start() with error returned: %u\n",
                       err);
            }
            uc_reg_read(uc, UC_ARM_REG_PC, &addr);
        }
    }

    uc_reg_read(uc, UC_ARM_REG_R2, &r2);
    uc_reg_read(uc, UC_ARM_REG_R3, &r3);

    uc_close(uc);

    *r2_out = r2;
    *r3_out = r3;
}

static void test_thumb_ite(void)
{
    uint32_t r2, r3;
    uint32_t step_r2, step_r3;

    printf("Emulate a THUMB ITE block as a whole or per instruction.\n");

    // Run once.
    printf("Running the entire binary.\n");
    test_thumb_ite_internal(false, &r2, &r3);
    printf(">>> R2: %d\n", r2);
    printf(">>> R3: %d\n\n", r3);

    // Step each instruction.
    printf("Running the binary one instruction at a time.\n");
    test_thumb_ite_internal(true, &step_r2, &step_r3);
    printf(">>> R2: %d\n", step_r2);
    printf(">>> R3: %d\n\n", step_r3);

    if (step_r2 != r2 || step_r3 != r3) {
        printf("Failed with ARM ITE blocks stepping!\n");
    }
}

static void test_read_sctlr(void)
{
    uc_engine *uc;
    uc_err err;
    uc_arm_cp_reg reg;

    printf("Read the SCTLR register.\n");

    err = uc_open(UC_ARCH_ARM, UC_MODE_ARM, &uc);
    if (err != UC_ERR_OK) {
        printf("Failed on uc_emu_start() with error returned: %u\n", err);
    }

    // SCTLR. See arm reference.
    reg.cp = 15;
    reg.is64 = 0;
    reg.sec = 0;
    reg.crn = 1;
    reg.crm = 0;
    reg.opc1 = 0;
    reg.opc2 = 0;

    err = uc_reg_read(uc, UC_ARM_REG_CP_REG, &reg);
    if (err != UC_ERR_OK) {
        printf("Failed on uc_reg_read() with error returned: %u\n", err);
    }

    printf(">>> SCTLR = 0x%" PRIx32 "\n", (uint32_t)reg.val);
    printf(">>> SCTLR.IE = %" PRId32 "\n", (uint32_t)((reg.val >> 31) & 1));
    printf(">>> SCTLR.B = %" PRId32 "\n", (uint32_t)((reg.val >> 7) & 1));

    uc_close(uc);
}

int main(int argc, char **argv, char **envp)
{
    test_arm();

    printf("==========================\n");
    test_thumb();

    printf("==========================\n");
    test_armeb();

    printf("==========================\n");
    test_thumbeb();

    printf("==========================\n");
    test_thumb_mrs();

    printf("==========================\n");
    test_thumb_ite();

    printf("==========================\n");
    test_read_sctlr();

    return 0;
}

File: bindings/java/src/test/java/samples/Sample_arm.java
/* Unicorn Emulator Engine */
/* By Nguyen Anh Quynh, 2015 */

/* Sample code to demonstrate how to emulate ARM code */

package samples;

import java.util.Arrays;

import unicorn.*;

public class Sample_arm implements UnicornConst, ArmConst {

    /** code to be emulated {@code mov r0, #0x37; sub r1, r2, r3} */
    // private static final byte[] ARM_CODE = Utils.hexToBytes("3700a0e3031042e0");
    /** code to be emulated {@code nop} */
    private static final byte[] ARM_CODE = Utils.hexToBytes("00f020e3");

    /** code to be emulated {@code sub sp, #0xc} */
    private static final byte[] THUMB_CODE = Utils.hexToBytes("83b0");

    /** code to be emulated
     * <pre>
     * cmp r2, r3
     * it ne
     * mov r2, #0x68
     * mov r2, #0x4d
     * </pre>
     */
    private static final byte[] ARM_THUMB_COND_CODE =
        Utils.hexToBytes("9a4214bf68224d22");

    /** code to be emulated {@code mov r0, #0x37; sub r1, r2, r3} */
    private static final byte[] ARM_CODE_EB =
        Utils.hexToBytes("e3a00037e0421003");
    /** code to be emulated {@code sub sp, #0xc} */
    private static final byte[] THUMB_CODE_EB = Utils.hexToBytes("b083");

    /** {@code 0xf3ef8014 - mrs r0, control} */
    private static final byte[] THUMB_CODE_MRS = Utils.hexToBytes("eff31480");

    /** memory address where emulation starts */
    private static final long ADDRESS = 0x10000;

    private static final BlockHook hook_block =
        (uc, address, size, user_data) -> {
            System.out.format(
                ">>> Tracing basic block at 0x%x, block size = 0x%x\n",
                address, size);
        };

    private static final CodeHook hook_code =
        (uc, address, size, user_data) -> {
            System.out.format(
                ">>> Tracing instruction at 0x%x, instruction size = 0x%x\n",
                address, size);
        };

    public static void test_arm() {
        long r0 = 0x1234L; // R0 register
        long r2 = 0x6789L; // R1 register
        long r3 = 0x3333L; // R2 register

        System.out.println("Emulate ARM code");

        // Initialize emulator in ARM mode
        Unicorn u = new Unicorn(UC_ARCH_ARM, UC_MODE_ARM);

        // map 2MB memory for this emulation
        u.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        // write machine code to be emulated to memory
        u.mem_write(ADDRESS, ARM_CODE);

        // initialize machine registers
        u.reg_write(UC_ARM_REG_R0, r0);
        u.reg_write(UC_ARM_REG_R2, r2);
        u.reg_write(UC_ARM_REG_R3, r3);

        // tracing all basic blocks with customized callback
        u.hook_add(hook_block, 1, 0, null);

        // tracing one instruction at ADDRESS with customized callback
        u.hook_add(hook_code, ADDRESS, ADDRESS, null);

        // emulate machine code in infinite time (last param = 0), or when
        // finishing all the code.
        u.emu_start(ADDRESS, ADDRESS + ARM_CODE.length, 0, 0);

        // now print out some registers
        System.out.println(">>> Emulation done. Below is the CPU context");

        System.out.format(">>> R0 = 0x%x\n", u.reg_read(UC_ARM_REG_R0));
        System.out.format(">>> R1 = 0x%x\n", u.reg_read(UC_ARM_REG_R1));
    }

    public static void test_thumb() {
        long sp = 0x1234L; // R0 register

        System.out.println("Emulate THUMB code");

        // Initialize emulator in ARM mode
        Unicorn u = new Unicorn(UC_ARCH_ARM, UC_MODE_THUMB);

        // map 2MB memory for this emulation
        u.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        // write machine code to be emulated to memory
        u.mem_write(ADDRESS, THUMB_CODE);

        // initialize machine registers
        u.reg_write(UC_ARM_REG_SP, sp);

        // tracing all basic blocks with customized callback
        u.hook_add(hook_block, 1, 0, null);

        // tracing one instruction at ADDRESS with customized callback
        u.hook_add(hook_code, ADDRESS, ADDRESS, null);

        // emulate machine code in infinite time (last param = 0), or when
        // finishing all the code.
        u.emu_start(ADDRESS | 1, ADDRESS + THUMB_CODE.length, 0, 0);

        // now print out some registers
        System.out.print(">>> Emulation done. Below is the CPU context\n");
        System.out.format(">>> SP = 0x%x\n", u.reg_read(UC_ARM_REG_SP));
    }

    public static void test_armeb() {
        long r0 = 0x1234L; // R0 register
        long r2 = 0x6789L; // R1 register
        long r3 = 0x3333L; // R2 register

        System.out.println("Emulate ARM Big-Endian code");

        // Initialize emulator in ARM mode
        Unicorn uc = new Unicorn(UC_ARCH_ARM, UC_MODE_ARM | UC_MODE_BIG_ENDIAN);

        // map 2MB memory for this emulation
        uc.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        // write machine code to be emulated to memory
        uc.mem_write(ADDRESS, ARM_CODE_EB);

        // initialize machine registers
        uc.reg_write(UC_ARM_REG_R0, r0);
        uc.reg_write(UC_ARM_REG_R2, r2);
        uc.reg_write(UC_ARM_REG_R3, r3);

        // tracing all basic blocks with customized callback
        uc.hook_add(hook_block, 1, 0, null);

        // tracing one instruction at ADDRESS with customized callback
        uc.hook_add(hook_code, ADDRESS, ADDRESS, null);

        // emulate machine code in infinite time (last param = 0), or when
        // finishing all the code.
        uc.emu_start(ADDRESS, ADDRESS + ARM_CODE_EB.length, 0, 0);

        // now print out some registers
        System.out.println(">>> Emulation done. Below is the CPU context");
        System.out.format(">>> R0 = 0x%x\n", uc.reg_read(UC_ARM_REG_R0));
        System.out.format(">>> R1 = 0x%x\n", uc.reg_read(UC_ARM_REG_R1));
    }

    public static void test_thumbeb() {
        long sp = 0x1234L;

        System.out.println("Emulate THUMB Big-Endian code");

        // Initialize emulator in ARM mode
        Unicorn uc =
            new Unicorn(UC_ARCH_ARM, UC_MODE_THUMB + UC_MODE_BIG_ENDIAN);

        // map 2MB memory for this emulation
        uc.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        // write machine code to be emulated to memory
        uc.mem_write(ADDRESS, THUMB_CODE_EB);

        // initialize machine registers
        uc.reg_write(UC_ARM_REG_SP, sp);

        // tracing all basic blocks with customized callback
        uc.hook_add(hook_block, 1, 0, null);

        // tracing one instruction at ADDRESS with customized callback
        uc.hook_add(hook_code, ADDRESS, ADDRESS, null);

        // emulate machine code in infinite time (last param = 0), or when
        // finishing all the code.
        // Note we start at ADDRESS | 1 to indicate THUMB mode.
        uc.emu_start(ADDRESS | 1, ADDRESS + THUMB_CODE_EB.length, 0, 0);

        // now print out some registers
        System.out.println(">>> Emulation done. Below is the CPU context");
        System.out.format(">>> SP = 0x%x\n", uc.reg_read(UC_ARM_REG_SP));
    }

    public static void test_thumb_mrs() {
        System.out.println("Emulate THUMB MRS instruction");
        // 0xf3ef8014 - mrs r0, control

        // Initialize emulator in ARM mode
        Unicorn uc = new Unicorn(UC_ARCH_ARM, UC_MODE_THUMB);

        // Setup the cpu model.
        uc.ctl_set_cpu_model(UC_CPU_ARM_CORTEX_M33);

        // map 2MB memory for this emulation
        uc.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        // write machine code to be emulated to memory
        uc.mem_write(ADDRESS, THUMB_CODE_MRS);

        // tracing all basic blocks with customized callback
        uc.hook_add(hook_block, 1, 0, null);

        // tracing one instruction at ADDRESS with customized callback
        uc.hook_add(hook_code, ADDRESS, ADDRESS, null);

        // emulate machine code in infinite time (last param = 0), or when
        // finishing all the code.

        // Note we start at ADDRESS | 1 to indicate THUMB mode.
        uc.emu_start(ADDRESS | 1, ADDRESS + THUMB_CODE_MRS.length, 0, 1);

        // now print out some registers
        System.out.println(">>> Emulation done. Below is the CPU context");

        long pc = uc.reg_read(UC_ARM_REG_PC);
        System.out.format(">>> PC = 0x%x\n", pc);
        if (pc != ADDRESS + 4) {
            System.out.format("Error, PC was 0x%x, expected was 0x%x.\n", pc,
                ADDRESS + 4);
        }
    }

    private static void test_thumb_ite_internal(boolean step, long[] r2r3) {
        Unicorn uc = new Unicorn(UC_ARCH_ARM, UC_MODE_THUMB);

        uc.mem_map(ADDRESS, 2 * 1024 * 1024, UC_PROT_ALL);

        uc.mem_write(ADDRESS, ARM_THUMB_COND_CODE);

        uc.reg_write(UC_ARM_REG_SP, 0x1234L);

        uc.reg_write(UC_ARM_REG_R2, 0);
        uc.reg_write(UC_ARM_REG_R3, 1);

        if (!step) {
            uc.emu_start(ADDRESS | 1, ADDRESS + ARM_THUMB_COND_CODE.length, 0,
                0);
        } else {
            long addr = ADDRESS;
            for (int i = 0; i < ARM_THUMB_COND_CODE.length / 2; i++) {
                uc.emu_start(addr | 1, ADDRESS + ARM_THUMB_COND_CODE.length, 0,
                    1);
                addr = uc.reg_read(UC_ARM_REG_PC);
            }
        }

        r2r3[0] = uc.reg_read(UC_ARM_REG_R2);
        r2r3[1] = uc.reg_read(UC_ARM_REG_R3);
    }

    public static void test_thumb_ite() {
        long[] r2r3 = new long[2];
        long[] step_r2r3 = new long[2];

        System.out.println(
            "Emulate a THUMB ITE block as a whole or per instruction.");

        // Run once.
        System.out.println("Running the entire binary.");
        test_thumb_ite_internal(false, r2r3);
        System.out.format(">>> R2: %d\n", r2r3[0]);
        System.out.format(">>> R3: %d\n\n", r2r3[1]);

        // Step each instruction.
        System.out.println("Running the binary one instruction at a time.");
        test_thumb_ite_internal(true, step_r2r3);
        System.out.format(">>> R2: %d\n", step_r2r3[0]);
        System.out.format(">>> R3: %d\n\n", step_r2r3[1]);

        if (!Arrays.equals(r2r3, step_r2r3)) {
            System.out.println("Failed with ARM ITE blocks stepping!");
        }
    }

    public static void test_read_sctlr() {
        System.out.println("Read the SCTLR register.");

        Unicorn uc = new Unicorn(UC_ARCH_ARM, UC_MODE_ARM);

        // SCTLR. See arm reference.
        Arm_CP reg = new Arm_CP(15, 0, 0, 1, 0, 0, 0);
        long val = (Long) uc.reg_read(UC_ARM_REG_CP_REG, reg);

        System.out.format(">>> SCTLR = 0x%x\n", val & 0xffffffffL);
        System.out.format(">>> SCTLR.IE = %d\n", (val >> 31) & 1);
        System.out.format(">>> SCTLR.B = %d\n", (val >> 7) & 1);
    }

    public static void main(String args[]) {
        test_arm();
        System.out.print("==========================\n");
        test_thumb();

        System.out.print("==========================\n");
        test_armeb();

        System.out.print("==========================\n");
        test_thumbeb();

        System.out.print("==========================\n");
        test_thumb_mrs();

        System.out.print("==========================\n");
        test_thumb_ite();

        System.out.print("==========================\n");
        test_read_sctlr();
    }

}
